### ISSUE TITLE
Full Module RNG Scope

### ISSUE THREAD

**Issue by joeryjoery** [google/flax#3506]
Converting this to an issue since I'm repeatedly stumbling into the current non-streamlined implementation for module random-key management (see discussion below; it doesn't present a solution yet, but it writes down some ideas) .

The biggest issue that the current design has is that it prevents me from writing composing generic modules without having to awkardly pass a scope of RNGs (and likely make a mistake). 

This snippet shows a example use-case for sampling and applying random functions,
```python
from typing import Generic, TypeVar

import jax
import jax.numpy as jnp

from flax import struct
from flax import linen as nn


M = TypeVar('M', bound=nn.Module)

@struct.dataclass
class Composite(Generic[M]):
    module: M

    def do(self, key: jax.Array, ins: jax.Array):
        # Sample random module parameters, and apply the function.
        
        key_init, keys_sample = jax.random.split(key)
        
        # Problem: Throws an error if an arbitrary key in `M` is not covered.
        params = self.module.init(key, ins)

        # Problem: I do not know the RNG scope of self.module
        outs = self.module.apply(params, ins)
        
        return outs


c = Composite(nn.Dense(5))
c.do(jax.random.key(0), jnp.ones(5))
```

### Discussed in https://github.com/google/flax/discussions/3375

<div type='discussions-op-text'>

<sup>Originally posted by **joeryjoery** September 28, 2023</sup>
Hi, if I implement a Module that makes use of `make_rng`, e.g., Dropout, I now have to manually specify the `rng` when I use the Module's call function. Though I completely agree that the RNG streams should be split into distinct paths like it is now, it is quite annoying that I have to bookkeep this at every call to `Module.apply`.

```python
import jax
import flax.linen as nn


rng = jax.random.PRNGKey(0)


class A(nn.Module):

    def __call__(self):
        x = self.make_rng('x')
        y = self.make_rng('y')

        return jax.random.uniform(x, ()) + jax.random.uniform(y, ())

myA = A()

# Currently
keys = jax.random.split(rng, 3)
scope = {'params', 'x', 'y'}  # I don't want to do this manually

# Desired functionality (or something like this)
# scope = myA.get_rng_scope()

rngs = dict(zip(scope, keys))

v = myA.init(rngs)
_ = myA.apply(v, rngs=rngs)
```
Is there already something like this implemented? Or could this be done in the future?

I'd also argue that flax should allow me to just pass in one global RNG key to either `init` or `apply`.
```python
# Even better (just use one key)
_ = myA.apply(v, rngs=jax.random.PRNGKey(0))
```
This functionality would be great to have with the *option* to pass in the RNGS explicitly. This has no consequences for reproducibility since the rngs need to be split somewhere anyway. 

At the moment the RNG design forces me to write work-arounds in my code which requires me to pass around a namespace that tracks all registered scopes. But this is quite error-prone and ideally this would be handled internally.</div>

**Comment by joeryjoery** [comment]
I'm not fully aware of the flax internals, I wasn't aware of `scope.rngs` so this is great to know! I can already use this to unit-test e.g., typos in creating rngs (e.g., I have the RNG stream `dropout` and somewhere else `drpout`).

Though in your code example you still have to pass in all keys explicitly in the init, which I might not know in advance for generic Modules and would have to explicitly trace.

It would be amazing if this would be handled internally.

**Comment by chiamp** [comment]
After #3669, you can now pass in a single RNG key under the `'default'` name to `.init` and `.apply`. For example `{'default': jax.random.key(0)}`

**Comment by joeryjoery** [comment]
> After #3669, you can now pass in a single RNG key under the `'default'` name to `.init` and `.apply`. For example `{'default': jax.random.key(0)}`

Great, this is a much better alternative. Thanks a lot!
